File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/ValidValuesString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class ValidValuesString {
    private static final String DNCString = "?DNC?";

    public String value;
    public String isValid;
    public String notes;

    public ValidValuesString(String value, String isValid, String notes) {
        this.value = value;
        this.isValid = isValid;
        this.notes = notes;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static ValidValuesString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 3, "ValidValues");
        return new ValidValuesString(parts.get(0), parts.get(1), parts.get(2));
    }

    public ValidValuesString(String text) {
        ValidValuesString parsed = fromText(text);
        this.value = parsed.value;
        this.isValid = parsed.isValid;
        this.notes = parsed.notes;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ValidValuesString)) return false;
        ValidValuesString that = (ValidValuesString) o;
        return (DNCString.equals(value) || DNCString.equals(that.value) || Objects.equals(value, that.value))
            && (DNCString.equals(isValid) || DNCString.equals(that.isValid) || Objects.equals(isValid, that.isValid))
            && (DNCString.equals(notes) || DNCString.equals(that.notes) || Objects.equals(notes, that.notes));
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, isValid, notes);
    }

    @Override
    public String toString() {
        return Tokens.token(value) + " " + Tokens.token(isValid) + " " + Tokens.token(notes);
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/ValidValuesString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class ValidValuesString
    {
        public string value;
        public string isValid;
        public string notes;

        public ValidValuesString(string value, string isValid, string notes)
        {
            this.value = value;
            this.isValid = isValid;
            this.notes = notes;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static ValidValuesString FromText(string text)
        {
            var parts = Tokens.Require(text, 3, "ValidValues");
            return new ValidValuesString(parts[0], parts[1], parts[2]);
        }

        public ValidValuesString(string text)
        {
            var parsed = FromText(text);
            this.value = parsed.value;
            this.isValid = parsed.isValid;
            this.notes = parsed.notes;
        }

        public ValidValuesTyped ToValidValuesTyped()
        {
            return new ValidValuesTyped(
                this.value,
                (this.isValid.Equals("true", System.StringComparison.OrdinalIgnoreCase) || this.isValid.Equals("t", System.StringComparison.OrdinalIgnoreCase) || this.isValid.Equals("yes", System.StringComparison.OrdinalIgnoreCase) || this.isValid.Equals("y", System.StringComparison.OrdinalIgnoreCase) || this.isValid == "1"),
                this.notes
            );
        }

        public override string ToString()
        {
            return Tokens.Token(value) + " " + Tokens.Token(isValid) + " " + Tokens.Token(notes);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not ValidValuesString other) return false;
            return (DNCString == this.value || DNCString == other.value || object.Equals(this.value, other.value))
                && (DNCString == this.isValid || DNCString == other.isValid || object.Equals(this.isValid, other.isValid))
                && (DNCString == this.notes || DNCString == other.notes || object.Equals(this.notes, other.notes));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.value);
            h.Add(this.isValid);
            h.Add(this.notes);
            return h.ToHashCode();
        }
    }
}
